src/particles/diagnostics/EmittanceInvariants.cpp

```diff
@@ -0,0 +1,141 @@
+/* Copyright 2022-2023 The Regents of the University of California, through Lawrence
+ *           Berkeley National Laboratory (subject to receipt of any required
+ *           approvals from the U.S. Dept. of Energy). All rights reserved.
+ *
+ * This file is part of ImpactX.
+ *
+ * Authors: Chad Mitchell, Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#include "CovarianceMatrixMath.H"
+
+#include <stdexcept>
+#include <AMReX_Extension.H>
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+#include <cmath>
+#include <vector>
+#include <tuple>
+
+
+namespace impactx::diagnostics
+{
+
+    /** This function returns the three independent kinetic invariants
+     *  denoted I2, I4, and I6 as constructed from the 6x6
+     *  beam covariance matrix.  These three quantities are invariant
+     *  under any linear symplectic transport map, and are used in the
+     *  calculation of the three eigenemittances.
+     *
+     * input - Sigma symmetric 6x6 covariance matrix
+     * returns - tuple containing invarants I2, I4, and I6
+     */
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    KineticInvariants (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> Sigma
+    )
+    {
+        using namespace amrex::literals;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> invariants;
+        amrex::ParticleReal I2 = 0.0_prt;
+        amrex::ParticleReal I4 = 0.0_prt;
+        amrex::ParticleReal I6 = 0.0_prt;
+
+        // Intermediate matrices used for storage.
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> S1;
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> S2;
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> S4;
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> S6;
+
+        // Construct the matrix S1 = Sigma*J.  This is a
+        // permutation of the columns of Sigma with
+        // a change of sign.
+        for (int i = 1; i < 7; i++) {
+            for (int j = 1; j < 7; j++) {
+                if (j % 2) {
```

Comment:
To be more explicit and avoiding an implicit int-to-bool cast, maybe:
```suggestion
                if (j % 2 == 1) {
```